src/set_distances.jl

```diff
@@ -0,0 +1,183 @@
+
+function _correct_distance_to_set(d::Real)
+    return ifelse(d < 0, zero(d), d)
+end
+
+function _correct_distance_to_set(d::AbstractVector{T}) where {T <: Real}
+    return [_correct_distance_to_set(di) for di in d]
+end
+
+"""
+    distance_to_set(v, s)
+
+Compute the distance of a value to a set.
+For some vector-valued sets, can return a vector of distances.
+When `v ∈ s`, the distance is zero (or all individual distances are zero).
+
+Each set `S` implements `distance_to_set(v::T, s::S)` with `T` of appropriate type.
+"""
+function distance_to_set end
+
+distance_to_set(v::Real, s::LessThan) = _correct_distance_to_set(v - s.upper)
+distance_to_set(v::Real, s::GreaterThan) = _correct_distance_to_set(s.lower - v)
+distance_to_set(v::Real, s::EqualTo) = abs(v - s.value)
+
+distance_to_set(v::T, s::Interval) where {T <: Real} = max(s.lower - v, v - s.upper, zero(T))
+
+function distance_to_set(v::AbstractVector{<:Real}, s::Reals)
+    length(v) != dimension(s) && throw(DimensionMismatch("Mismatch between value and set"))
+    return zeros(length(v))
+end
+
+function distance_to_set(v::AbstractVector{<:Real}, s::Zeros)
+    length(v) != dimension(s) && throw(DimensionMismatch("Mismatch between value and set"))
+    return abs.(v)
+end
+
+function distance_to_set(v::AbstractVector{<:Real}, s::Nonnegatives)
+    length(v) != dimension(s) && throw(DimensionMismatch("Mismatch between value and set"))
+    return _correct_distance_to_set(-v)
```

Comment:
```Julia
function distance_to_set(x::AbstractVector{T}, s::Nonnegatives) where {T<:Real}
    _check_dimension(x, s)
    return max.(zero(T), -x)
end

Reply:
we need consistency somewhere in the decision. Either calling helper functions is allowed or we re-write everything in the function body. Why having all the `max(zero(T), d)` in the function body but add an indirection for the `_check_dimension`?

Reply:
The `max(0, -x)` is the definition of the distance, so it should be in the function body.

`_check_dimension` is not part of the definition of distance, and is just to help users from making mistakes. There isn't a need for it to be in the function body explicitly.